core/src/main/java/io/parsingdata/metal/Shorthand.java

```diff
@@ -189,6 +189,15 @@ private Shorthand() {}
     /** @see Until */ public static Token until(final String name, final Token terminator, final Encoding encoding) { return until(name, null, terminator, encoding); }
     /** @see Until */ public static Token until(final String name, final Token terminator) { return until(name, terminator, null); }
 
+    public static Token defU(final String name, final ValueExpression initialSize, final ValueExpression stepSize, final ValueExpression maxSize, final Token terminator, final Encoding encoding) { return new Until(name, initialSize, stepSize, maxSize, terminator, false, encoding); }
```

Comment:
We do not like the name `defU`, U meaning Unterminated, anymore. Mostly, because the def is not unterminated, it is terminated, but we just do not know where yet. We are considering the following names:
* tdefEx: terminated def excluding terminator. this allows for the equivalent `tdefIn` for which we want to rename the original until shorthand to (and keeping the original name for backward compatibility)
* defn: in same line of repn, although this can be confusing, since repn uses a valueexpression to determine how far it should go, while rep is dynamic. so then def should actually be defn and this new token is def... let's not go there.

Reply:
> ["There are only two hard things in Computer Science: cache invalidation and naming things"](https://skeptics.stackexchange.com/questions/19836/has-phil-karlton-ever-said-there-are-only-two-hard-things-in-computer-science)

I like your first suggestion: `tdef`.

As a side note: Kaitai has also encountered this exact case, and they have [the special keyword `include`](https://doc.kaitai.io/user_guide.html#delimited-struct-advanced) that indicates whether a configured `terminator` should be part of the parsed value or not.

Reply:
I like the idea of appending `Ex` or `In`, that makes it really clear what the method does. So my vote goes to `tdexIn` and `tdefEx`.

Reply:
![defu](https://jvdb.org/defu.png)

Reply:
The confusing part of the current `until` implementation is that it is actually a combination token that parses two defs. The proposed alternative implementation, `defU`, only parses a single `def`, which is why we would like to let the shorthand of this variant also start with `def`.

In addition, we are able to write `until` behaviour using `defU` icw `seq` and `post`, so we could decide to let the `Until` class always set `includeTerminator` to false and if a terminator is desired to be included, we can resolve that with shorthands only.

I would therefor propose to rename `Until.java` to `DefUntil.java` (or `DefBoundedByToken.java`) as is will be implemented to parse a single def until what ever token was defined and then define the following shorthands from the user's perspective:
* def(name, token): (an overload method of the current def) the specified terminator should not be part of the resulting def and should not be part of the parsegraph. the token simply defines the boundary of the def.
* defIn(name, token): the specified terminator should be part of the resulting def, e.g. line + \n
* defs(name, token): a shorthand for seq(def(name, token), token), such that the terminator is also parsed and part of the parsegraph.

I think we then cover all the possible until variants.

Reply:
I think calling them all `def` is perhaps a good idea after all. Essentially it retains this idea that there are basically two types of `Token`s: combinators and value definitions. The latter is then always called `Def`, regardless of how it is defined -- as either something with a provided size, a provided terminator or through some provided `Expression` that does whatever it needs to determine whether a growing candidate value is "complete".

Reply:
Agreed. For the longest time we only had `def` as our value definition in Metal. Until `until` came along (pun intended). Ever since then we've had to explain this whenever we've shown Metal to new team members.